Patches/Specular.cpp

```diff
@@ -0,0 +1,396 @@
+	float diffuseG;
+	float diffuseB;
+	float ambientX;
+	float ambientR;
+	float ambientG;
+	float ambientB;
+	float specularX;
+	float specularR;
+	float specularG;
+	float specularB;
+	int reserved3;
+	int unkIndex;
+	int primCount;
+	int reserved4;
+};
+
+static LightSource fakeLight = { {D3DLIGHT_DIRECTIONAL} };
+static bool useFakeLight = false;
+static bool inSpecialLightZone = false;
+static int fakeLightIndex = -1;
+static int materialCount = 0;
+static ModelMaterial* pFirstMaterial = nullptr;
+static ModelMaterial* pCurrentMaterial = nullptr;
+
+ModelID(__cdecl* GetModelID)() = nullptr;
+static int(__cdecl* GetLightSourceCount)() = nullptr;
+static LightSource* (__cdecl* GetLightSourceStruct)(int) = nullptr;
+static void(__cdecl* ActorDrawOpaque)(ModelMaterial*) = nullptr;
+static void(__cdecl* DoActorOpaqueStuff)(ModelOffsetTable*, void*) = nullptr;
+static IDirect3DDevice8** pD3DDevice = nullptr;
+
+static int GetCurrentMaterialIndex()
+{
+	int index = 0;
+	ModelMaterial* pCursor = pFirstMaterial;
+
+	while (index < materialCount)
+	{
+		if (pCurrentMaterial == pCursor)
+			return index;
+
+		pCursor = reinterpret_cast<ModelMaterial*>((reinterpret_cast<char*>(pCursor) + pCursor->materialLength));
+		index++;
+	}
+
+	return -1;
+}
+
+static bool IsJames(ModelID id)
+{
+	switch (id)
+	{
+	case ModelID::chr_jms_lll_jms:
+	case ModelID::chr_jms_hll_jms:
+	case ModelID::chr_jms_hhl_jms:
+	case ModelID::chr_jms_hhh_jms:
+	case ModelID::chr_jms_rlll_jms:
+	case ModelID::chr_jms_rhll_jms:
+	case ModelID::chr_jms_rhhl_jms:
+	case ModelID::chr_jms_rhhh_jms:
+		return true;
+	}
+
+	return false;
+}
+
+static bool IsMariaExcludingEyes(ModelID id)
+{
+	switch (id)
+	{
+	case ModelID::chr_mar_lll_mar:
+	case ModelID::chr_mar_hhh_mar:
+	case ModelID::chr2_mar_lxx_mar:
+	case ModelID::chr_mar_rlll_mar:
+	case ModelID::chr_mar_rhhh_mar:
+	case ModelID::chr2_mar_rlxx_mar:
+		if (GetCurrentMaterialIndex() != 3) return true;
+		break;
+	case ModelID::chr_item_dmr:
+		if (GetCurrentMaterialIndex() != 1) return true;
+		break;
+	}
+
+	return false;
+}
+
+static bool IsMariaEyes(ModelID id)
+{
+	switch (id)
+	{
+	case ModelID::chr_mar_lll_mar:
+	case ModelID::chr_mar_hhh_mar:
+	case ModelID::chr2_mar_lxx_mar:
+	case ModelID::chr_mar_rlll_mar:
+	case ModelID::chr_mar_rhhh_mar:
+	case ModelID::chr2_mar_rlxx_mar:
+		if (GetCurrentMaterialIndex() == 3) return true;
+		break;
+	case ModelID::chr_item_dmr:
+		if (GetCurrentMaterialIndex() == 1) return true;
+		break;
+	}
+
+	return false;
+}
+
+static void __cdecl Part0(ModelOffsetTable* pOffsetTable, void* arg2)
+{
+	// This function replaces a call to `void DoActorOpaqueStuff(ModelOffsetTable* pOffsetTable, void* arg2)`
+	// Backup the materialCount and pointer to first material for use in later Parts
+
+	materialCount = pOffsetTable->materialCount;
+	pFirstMaterial = (ModelMaterial*)((char*)pOffsetTable + pOffsetTable->materialsOffset);
+
+	DoActorOpaqueStuff(pOffsetTable, arg2);
+}
+
+static int __cdecl Part1()
```

Comment:
Is there a reason this function is not in the "IDirect3DDevice8.cpp" file? It should be part of the GetLightSourceCount() function in that file.  Can this get a better name than `Part1`?

Reply:
This isn't actually a call to any D3D function. A `LightSource` isn't a `D3DLIGHT8`, though it contains one. My choice of name probably doesn't make that clear, I could rename `LightSource` `SH2LightSource` or something to help clarify?

Reply:
I see.  Ok then you can keep it as is.